clippy_lints/src/methods/unnecessary_iter_cloned.rs

```diff
@@ -40,6 +40,63 @@ pub fn check_for_loop_iter(
         && !clone_or_copy_needed
         && let Some(receiver_snippet) = snippet_opt(cx, receiver.span)
     {
+        // Issue 12098
+        // https://github.com/rust-lang/rust-clippy/issues/12098
+        // if the assignee have `mut borrow` conflict with the iteratee
+        // the lint should not execute, former didn't consider the mut case
+
+        // check whether `expr` is mutable
+        fn is_mutable(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+            if let Some(hir_id) = path_to_local(expr)
+                && let Node::Pat(pat) = cx.tcx.hir_node(hir_id)
+            {
+                matches!(pat.kind, PatKind::Binding(BindingAnnotation::MUT, ..))
+            } else {
+                true
+            }
+        }
+
+        fn is_caller_or_fields_changed(cx: &LateContext<'_>, body: &Expr<'_>, caller: &Expr<'_>) -> bool {
+            // local private `walk_blocks` function
+            // to iterate all the expr in the block and subblocks
+            fn walk_blocks(cx: &LateContext<'_>, block: &Block<'_>, caller: &Expr<'_>) -> bool {
+                for stmt in block.stmts {
+                    let changed = match stmt.kind {
+                        StmtKind::Expr(e) | StmtKind::Semi(e) => match e.kind {
+                            ExprKind::Assign(assignee, _, _) | ExprKind::AssignOp(_, assignee, _) => {
+                                !can_mut_borrow_both(cx, caller, assignee)
+                            },
+                            ExprKind::Block(block, ..) => walk_blocks(cx, block, caller),
+                            _ => false,
+                        },
+                        _ => false,
+                    };
+                    if changed {
+                        return true;
+                    }
+                }
```

Comment:
I'm pretty sure, that this function will not work for expressions that nest blocks, like `if` expressions loops etc. I would highly recommend to use [`for_each_expr`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/visitors/fn.for_each_expr.html) over implmeneting a custom recursion.